### PR TITLE
Proper timeouts for jobs

### PR DESCRIPTION
**Pull request by ubuntolog**
It is supposed to close the issue https://github.com/EUDAT-GEF/GEF/issues/92

### CODE REVIEW

File: gefserver/config.json [hunk 1]
```diff
@@ -39,6 +39,7 @@
 		"VolumeInspection": 1000,
 		"FileDownload": 1000,
 		"Preparation": 100,
-		"JobExecution": 1000
+		"JobExecution": 100,
```
Comment (emanueldima) [inline gefserver/config.json]: This value should be higher by default (2 hours?), we can change it programmatically to test the feature.

File: gefserver/pier/pier.go [hunk 2]
```diff
@@ -146,6 +150,42 @@ func (p *Pier) BuildService(buildDir string) (db.Service, error) {
 	return service, nil
 }
 
+// startTimeOutTicker starts a clock that checks if a job exceeds an execution timeout
+func (p *Pier) startTimeOutTicker(jobId db.JobID) {
+	ticker := time.NewTicker(time.Second * time.Duration(p.checkInterval))
+	go func() {
```
Comment (emanueldima) [inline gefserver/pier/pier.go]: this goroutine is not necessary, there's another one when you're calling the startTimeOutTicker function

File: gefserver/pier/pier.go [hunk 3]
```diff
@@ -146,6 +150,42 @@ func (p *Pier) BuildService(buildDir string) (db.Service, error) {
...
+			durationTime := time.Duration(currentTime.Sub(startingTime))
+
+			if durationTime.Seconds() >= float64(p.jobExecTimeOut) {
+				p.db.SetJobState(job.ID, db.NewJobStateError("Job execution timeout exceeded", 1))
+				ticker.Stop()
+
+				theLastContainer := job.Tasks[len(job.Tasks)-1].ContainerID
+				_, err = p.docker.client.WaitContainerOrSwarmService(string(theLastContainer), true)
+				if err != nil {
+					p.db.SetJobState(job.ID, db.NewJobStateError("Container removal failed", 1))
```
Comment (emanueldima) [inline gefserver/pier/pier.go]: This error would overwrite the previous error, set it as 'Job execution timeout exceeded and container removal failed'

File: gefserver/main.go [hunk 4]
```diff
@@ -31,7 +31,7 @@ func main() {
 	defer d.Close()
 
 	var p *pier.Pier
-	p, err = pier.NewPier(&d, config.TmpDir)
+	p, err = pier.NewPier(&d, config.TmpDir, config.Timeouts.JobExecution, config.Timeouts.CheckInterval)
```
Comment (emanueldima) [inline gefserver/main.go]: It's too fine grained to send all timeouts individually, let's store the entire timeout structure into pier

File: gefserver/pier/pier.go [hunk 5]
```diff
@@ -146,6 +150,42 @@ func (p *Pier) BuildService(buildDir string) (db.Service, error) {
...
+				ticker.Stop()
+				break
+			}
+			if job.State.Code != -1 {
+				ticker.Stop()
+				break
+			}
+
+			startingTime := job.Created
+			currentTime := time.Now().UTC()
```
Comment (emanueldima) [inline gefserver/pier/pier.go]: this one is UTC, but all the ones we create in the db are local times; we should make an issue to transform them all

File: gefserver/pier/pier.go [hunk 6]
```diff
@@ -146,6 +148,40 @@ func (p *Pier) BuildService(buildDir string) (db.Service, error) {
...
 }
 
+// startTimeOutTicker starts a clock that checks if a job exceeds an execution timeout
+func (p *Pier) startTimeOutTicker(jobId db.JobID) {
+	ticker := time.NewTicker(time.Second * time.Duration(p.timeOuts.CheckInterval))
+	for range ticker.C {
+		job, err := p.db.GetJob(jobId)
+
+		if err != nil {
+			p.db.SetJobState(job.ID, db.NewJobStateError("Cannot get information about the job running", 1))
```
Comment (emanueldima) [inline gefserver/pier/pier.go]: check error returned by SetJobState, make a log entry on error

File: gefserver/pier/pier.go [hunk 7]
```diff
@@ -146,6 +148,40 @@ func (p *Pier) BuildService(buildDir string) (db.Service, error) {
...
+			ticker.Stop()
+			break
+		}
+
+		startingTime := job.Created
+		currentTime := time.Now()
+		durationTime := time.Duration(currentTime.Sub(startingTime))
+
+		if durationTime.Seconds() >= float64(p.timeOuts.JobExecution) {
+			p.db.SetJobState(job.ID, db.NewJobStateError("Job execution timeout exceeded", 1))
```
Comment (emanueldima) [inline gefserver/pier/pier.go]: check error returned by SetJobState, make a log entry on error

File: gefserver/pier/pier.go [hunk 8]
```diff
@@ -146,6 +148,40 @@ func (p *Pier) BuildService(buildDir string) (db.Service, error) {
...
+			ticker.Stop()
+
+			theLastContainer := job.Tasks[len(job.Tasks)-1].ContainerID
+			_, err = p.docker.client.WaitContainerOrSwarmService(string(theLastContainer), true)
+			if err != nil {
+				p.db.SetJobState(job.ID, db.NewJobStateError("Job execution timeout exceeded and container removal failed", 1))
```
Comment (emanueldima) [inline gefserver/pier/pier.go]: check error returned by SetJobState, make a log entry on error

**Comment by ubuntolog** [comment]
We also did not check errors in "runJob" function

**Comment by emanueldima** [comment]
What about tests?

**Comment by ubuntolog** [comment]
You mean to test the case when a job times out?

**Comment by emanueldima** [comment]
Yes, all functionality should be tested. Make an issue?

**Comment by ubuntolog** [comment]
I prefer to write a test instead

**Comment by ubuntolog** [comment]
I've added tests for timeouts. Please have a look

File: gefserver/tests/pier_test.go [hunk 9]
```diff
@@ -144,15 +144,61 @@ func TestExecution(t *testing.T) {
...
+	checkMsg(t, err, "reading config files")
+	config.Timeouts.JobExecution = 3  // forcing a small time out
+	config.Timeouts.CheckInterval = 2 // forcing a small check interval
+
+	db, dbfile, err := db.InitDbForTesting()
+	checkMsg(t, err, "creating db")
+	defer db.Close()
+	defer os.Remove(dbfile)
+
+	addUser(t, db, name1, email1)
```
Comment (emanueldima) [inline gefserver/tests/pier_test.go]: It's not needed to add users for this test.

File: gefserver/pier/internal/dckr/dckr.go [hunk 10]
```diff
@@ -371,14 +371,22 @@ func (c Client) getSwarmServiceIDByContainerID(containerID string) (string, erro
 }
 
 // StartImage takes a docker image, creates a container and starts it
-func (c Client) StartImage(id string, repoTag string, cmdArgs []string, binds []VolBind, limits def.LimitConfig, preparationTimeOut int64, executionTimeOut int64) (ContainerID, *bytes.Buffer, error) {
+func (c Client) StartImage(id string, repoTag string, cmdArgs []string, binds []VolBind, limits def.LimitConfig, preparationTimeOut float64, executionTimeOut float64) (ContainerID, *bytes.Buffer, error) {
```
Comment (emanueldima) [inline gefserver/pier/internal/dckr/dckr.go]: why float?

**Pull request merged**